src/connectors/impls/logging.rs

```diff
@@ -0,0 +1,137 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::connectors::prelude::*;
+use async_broadcast::{Receiver, RecvError};
+use tremor_pipeline::{LoggingMsg, LOGGING_CHANNEL};
+
+use crate::connectors::{prelude::KillSwitch, ConnectorBuilder, ConnectorType};
+
+#[derive()]
+pub(crate) struct PluggableLoggingConnector {
+    rx: Receiver<LoggingMsg>,
+}
+
+impl PluggableLoggingConnector {
+    pub(crate) fn new() -> Self {
+        Self {
+            rx: LOGGING_CHANNEL.rx(),
+        }
+    }
+}
+
+#[derive(Debug, Default)]
+pub(crate) struct Builder {}
+
+#[async_trait::async_trait]
+impl ConnectorBuilder for Builder {
+    fn connector_type(&self) -> ConnectorType {
+        "logs".into()
+    }
+    async fn build(
+        &self,
+        _id: &Alias,
+        _config: &ConnectorConfig,
+        _kill_switch: &KillSwitch,
+    ) -> Result<Box<dyn Connector>> {
+        Ok(Box::new(PluggableLoggingConnector::new()))
+    }
+}
+#[async_trait::async_trait()]
+impl Connector for PluggableLoggingConnector {
+    async fn connect(&mut self, _ctx: &ConnectorContext, _attempt: &Attempt) -> Result<bool> {
+        Ok(true)
+    }
+
+    async fn create_source(
+        &mut self,
+        source_context: SourceContext,
+        builder: SourceManagerBuilder,
+    ) -> Result<Option<SourceAddr>> {
+        let source = LoggingSource::new(self.rx.clone());
+        let addr = builder.spawn(source, source_context)?;
+        Ok(Some(addr))
+    }
+
+    async fn create_sink(
+        &mut self,
+        _sink_context: SinkContext,
+        _builder: SinkManagerBuilder,
+    ) -> Result<Option<SinkAddr>> {
+        Ok(None)
+    }
+    fn codec_requirements(&self) -> CodecReq {
+        CodecReq::Structured
+    }
+}
+pub(crate) struct LoggingSource {
+    rx: Receiver<LoggingMsg>,
+    origin_uri: EventOriginUri,
+}
+
+impl LoggingSource {
+    pub(crate) fn new(rx: Receiver<LoggingMsg>) -> Self {
+        Self {
+            rx,
+            origin_uri: EventOriginUri {
+                scheme: "tremor-logs".to_string(),
+                host: hostname(),
+                port: None,
+                path: vec![],
+            },
+        }
+    }
+}
+
+#[async_trait::async_trait()]
+impl Source for LoggingSource {
+    async fn pull_data(&mut self, _pull_id: &mut u64, _ctx: &SourceContext) -> Result<SourceReply> {
+        loop {
+            match self.rx.recv().await {
+                Ok(msg) => {
+                    break Ok(SourceReply::Structured {
+                        payload: msg.payload,
+                        origin_uri: msg.origin_uri.unwrap_or_else(|| self.origin_uri.clone()),
+                        stream: DEFAULT_STREAM_ID,
+                        port: None,
```

Comment:
Not a change request but perhaps we can use port for levels? perhaps we can add a TODO comment for this?

Reply:
I think it is not possible to make this change (I tried it) because this field "port" from `SourceReply` has a default value of "out" (modulo wrappers like `Option`, `Cow`, etc...) and is already used.
When launching the test "pluggable_logging_operator" with the tremor server running, there is just no output.

Reply:
All good a `TODO` around it so we remember to investigate later is good enough :)

Reply:
perfect 👍